src/v/v8_engine/executor.cc

```diff
@@ -0,0 +1,145 @@
+/*
+ * Copyright 2021 Vectorized, Inc.
+ *
+ * Licensed as a Redpanda Enterprise file under the Redpanda Community
+ * License (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ * https://github.com/vectorizedio/redpanda/blob/master/licenses/rcl.md
+ */
+
+#include "v8_engine/executor.h"
+
+#include "seastarx.h"
+#include "utils/mutex.h"
+#include "vassert.h"
+
+#include <seastar/core/alien.hh>
+#include <seastar/core/future.hh>
+#include <seastar/core/smp.hh>
+
+#include <exception>
+
+namespace v8_engine {
+
+namespace internal {
+
+// spsc_queue
+
+spsc_queue::spsc_queue(size_t queue_size)
+  : _items(queue_size) {}
+
+void spsc_queue::close() {
+    _is_stopped = true;
+    _push_mutex.broken();
+    _has_element_cv.notify_all();
+}
+
+ss::future<> spsc_queue::push(work_item* item) {
+    // Only one coroutine can wait on ss::readable_eventfd, so we need to use
+    // mutex for push
+    auto lock = co_await _push_mutex.get_units();
+
+    while (!_items.push(item)) {
+        co_await _is_not_full.wait();
+    }
+
+    _has_element_cv.notify_one();
+}
+
+work_item* spsc_queue::pop() {
+    work_item* item = nullptr;
+    std::unique_lock lock{_std_mutex};
+
+    // We need to use wait_for, because std::thread can miss notification from
+    // seastar thread (between check pred and sleep)
+    _has_element_cv.wait_for(lock, _timeout_cond_wait_std_thread_ms, [this] {
+        return !empty() || _is_stopped;
+    });
+
+    if (!empty()) {
+        bool need_notify = _items.write_available() == 0;
```

Comment:
Boost lock-free documentation says `Thread-safe and wait-free, should only be called from the producer thread`. This is a consumer thread.

Also, imagine a situation when this function returns 0. Right after that you're calling `pop` and the vacant space appears in the queue. But you're not sending the notification because the queue was full right before that.

I think that you can assume that `need_notify` is always true since you popping up the element out of the queue if it's not empty.

Reply:
![image](https://user-images.githubusercontent.com/18518081/126621784-bdcde8da-1a54-47af-996d-739dd8127350.png)

Reply:
> But you're not sending the notification because the queue was full right before that

In this case I will send notification

Reply:
I'm looking at this - https://www.boost.org/doc/libs/1_74_0/doc/html/boost/lockfree/spsc_queue.html#id-1_3_23_6_4_1_1_1_7_15-bb it says the opposite.

Reply:
Oh, I used old version( Thank you!

Reply:
delete this check